sdk/src/main/java/com/horizen/utils/MerklePath.java

```diff
@@ -112,4 +112,14 @@ public boolean isRightmost() {
         }
         return true;
     }
+
```

Comment:
Instead of defining and using `MerkleTreeUtils` object and quite complex  `SidechainHashMap.fullMerkleTree()` method: you can define `isRightmost` method here in the following way:
```
public boolean isRightmost(byte[] leaf) {
        byte[] currentNode = leaf;
        for(Pair<Byte, byte[]> neighbour : _merklePath) {
            if(neighbour.getKey() == (byte)0) {
                // left neighbour
                // concatenate neighbour value LEFT to currentNode
                currentNode = BytesUtils.reverseBytes(Utils.doubleSHA256HashOfConcatenation(BytesUtils.reverseBytes(neighbour.getValue()), BytesUtils.reverseBytes(currentNode)));
            } else { 
                // right neighbour
                // if not equal, merkle path doesn't lead to the rightmost leaf.
                if(!Arrays.equals(currentNode, neighbour.getValue()))
                    return false;
                // concatenate neighbour value RIGHT to currentNode
                currentNode = BytesUtils.reverseBytes(Utils.doubleSHA256HashOfConcatenation(BytesUtils.reverseBytes(currentNode), BytesUtils.reverseBytes(neighbour.getValue())));
            }
        }
        return true;
    }
```
In MainchainBlockReference during no proof verification, in a case of single rightmost neighbour, you will check:
1) merkle path of neighbour leads to proper root.
2) `leftNeighbour.merklePath.isRightmost(leftNeighbour.leafValue)` returns true.

Example schemas:
16 leaves:
```
                               R
                /                          \
             C1                               C2
        /           \                   /            \
      B1              B2              B3              B4
    /    \          /    \          /    \          /    \
  A1      A2      A3      A4      A5      A6      A7      A8
 / \     / \     / \     / \     / \     / \     / \     / \
1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16
```
11 leaves (mutated tree):
```

                               R
                /                          \
             C1                               C2
        /           \                   /            \
      B1              B2              B3              B3
    /    \          /    \          /    \          /    \
  A1      A2      A3      A4      A5      A6      A5      A6
 / \     / \     / \     / \     / \     / \     / \     / \
1   2   3   4   5   6   7   8   9   10  11  11  9  10  11  11
```

Reply:
As for me methods above is much easier for testing and using.
Moreover it's resistent to attack with Block data substitution.
In one of the next PR, Block is separated into block and data. Also MC Block Ref into Header and Data.
And MCRefData (where proofs and proofs of no data is present) doesn't have an impact on SidechainBlock Id.
So MCRefData can be substituted by a malicious actor.
Attack flow:
1. Honest SidechainBlock appeard. Block contains MCRef with no info about current Sidechain, so will have only proof of no data. Proof of know data contains only left neighbour (that must be the tightmost). Index of such neighbour is 5.
2. Malicious actor will take the block and substitute the merkle path to left neighbour (with index 7 in full merkle tree) to index 5 (in full or original merkle tree, doesn't matter).
3. Malicious actor will send broken block.
4. Any honest node, will invalidate malicious block with "invalid data" exception -> block id will be banned -> network can be splitted.